library/ssl_tls13_server.c

```diff
@@ -264,6 +264,32 @@ static int ssl_tls13_offered_psks_check_identity_match(
             ssl, identity, identity_len, obfuscated_ticket_age,
             session ) == SSL_TLS1_3_OFFERED_PSK_MATCH )
     {
+        /* RFC 8446 section 4.2.9
+         *
+         * Servers SHOULD NOT send NewSessionTicket with tickets that are not
+         * compatible with the advertised modes; however, if a server does so,
+         * the impact will just be that the client's attempts at resumption fail.
+         *
+         * We regard the ticket with incompatible key exchange modes as not match.
+         */
+        int kex_mode = 0;
+        if( session->ticket_flags &
+                MBEDTLS_SSL_TLS1_3_TICKET_ALLOW_PSK_RESUMPTION )
+        {
+            kex_mode |= MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_PSK;
+        }
+        if( session->ticket_flags &
+                MBEDTLS_SSL_TLS1_3_TICKET_ALLOW_PSK_EPHEMERAL_RESUMPTION )
+        {
+            kex_mode |= MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_PSK_EPHEMERAL;
+        }
+        MBEDTLS_SSL_DEBUG_TICKET_FLAGS( 4, session->ticket_flags );
+        if( mbedtls_ssl_tls13_check_kex_modes( ssl, kex_mode ) )
+        {
+            MBEDTLS_SSL_DEBUG_MSG( 3, ( "No suitable key exchange mode" ) );
+            return ( SSL_TLS1_3_OFFERED_PSK_NOT_MATCH );
+        }
+
```

Comment:
I agree that it could be put in `mbedtls_ssl_ticket_parse`. But there are two problems need to be addressed:
1. There is no ssl context in that callback. We couldn't know what key exchange modes advertised by client.
2. If the check failed, what error code should we return.
Any suggestions?

Reply:
> I agree that it could be put in `mbedtls_ssl_ticket_parse`. But there are two problems need to be addressed:
> 
>     1. There is no ssl context in that callback. We couldn't know what key exchange modes advertised by client.
That's a problem. let's do it outside firstly.
 
>     2. If the check failed, what error code should we return.

create a new error.  like `MBEDTLS_ERR_SSL_TICKET_INVALID_KEX_MODE`